src/vm/lib/metadata/agent.js

```diff
@@ -269,65 +269,76 @@ function checkStaleSocket(conn, opts, callback) {
-                    old_sockstat: conn.sockstat
+                    old_sockstat: zoneConn.sockstat
                 }, 'change in sockpath fs.stat signature: stale');
                 callback(null, true); // stale
                 return;
             }
         }
-        if ((conn.sockstat.ctime
-            && conn.sockstat.ctime.getTime()) !== stats.ctime.getTime()) {
+        if ((zoneConn.sockstat.ctime
+            && zoneConn.sockstat.ctime.getTime()) !== stats.ctime.getTime()) {
             opts.log.debug({
                 field: 'ctime',
-                sockpath: conn.sockpath,
+                sockpath: zoneConn.sockpath,
                 new_sockstat: stats,
-                old_sockstat: conn.sockstat
+                old_sockstat: zoneConn.sockstat
             }, 'change in sockpath fs.stat signature: stale');
             callback(null, true); // stale
             return;
         }
 
         // no error in stat, and no diff fields means exists: not stale
         opts.log.trace({
-            sockpath: conn.sockpath
+            sockpath: zoneConn.sockpath
         }, 'sockpath still exists and fs.stat signature matches: not stale');
         callback(null, false);
         return;
     });
 }
 
 /*
- * This function does an fs.fstat() on the 'fd' argument and attaches the
- * fs.Stats result to the 'conn' object as .sockstat. If there is an error
- * with fs.fstat() conn.sockstat will be set to an empty object.
+ * This function does an fs.stat() on the 'sockpath' of the zoneConn and
+ * attaches the fs.Stats result to the 'zoneConn' object as .sockstat. If there
+ * is an error with fs.stat() conn.sockstat will be set to an empty object.
  *
- * After the fstat has completed, callback() will be called. Any error from
- * fs.fstat() will be passed as the first and only argument to callback().
+ * After the stat has completed, callback() will be called. Any error from
+ * fs.stat() will be passed as the first and only argument to callback().
  */
-function addConnSockStat(fd, conn, callback) {
-    assert.number(fd, 'fd');
-    assert.object(conn, 'conn');
+function addConnSockStat(zoneConn, callback) {
+    assert.object(zoneConn, 'zoneConn');
+    assert.string(zoneConn.sockpath, 'zoneConn.sockpath');
     assert.func(callback, 'callback');
 
-    fs.fstat(fd, function _statSock(e, st) {
+    fs.stat(zoneConn.sockpath, function _statSock(e, st) {
         if (e) {
             // If there was an error w/ the stat, it's most likely because
             // the state of the world has changed. We'll fill in sockstat
             // with an empty object here so that checkStaleSocket will report
             // this as stale.
-            conn.sockstat = {};
+            zoneConn.sockstat = {};
         } else {
-            conn.sockstat = st;
+            zoneConn.sockstat = st;
         }
         callback(e);
     });
 }
 
+function closeZoneConnection(zoneConn) {
+    // .serverSocket is a net.Server for non-KVM
```

Comment:
I think we should assert that `zoneConn` is an object. I would also suggest asserting that one but only one of `close` or `destroy` methods got called.